### PR TITLE
Hybrid simulation model

### PR DESCRIPTION
**Pull request by DomNelson**
Implements a working (yet statistically untested) hybrid model. It extends the existing DTWF model with an extra parameter, the number of DTWF generations to be simulated before switching to the coalescent.

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Did a quick bit of investigative work, and the approach outlined above will work just fine. Will keep you posted.

**Comment by jeromekelleher** [comment]
OK, I've done some work here and I think it's all after coming together quite nicely. We can now do the following:
```python
    sim = msprime.simulator_factory(
        sample_size=10, model=msprime.DiscreteTimeWrightFisher(1000),
        random_generator=msprime.RandomGenerator(123))

    sim.run(100)
    ts = sim.get_tree_sequence()

    tree = ts.first()
    print(tree.draw(format="unicode"))
    print(ts.tables.nodes)

    sim.set_model("hudson", 1000)
    sim.run()
    ts = sim.get_tree_sequence()

    tree = ts.first()
    print(tree.draw(format="unicode"))
    print(ts.tables.nodes)
```
gives
```
               14        
              ┏┻┓        
    13        ┃ ┃        
 ┏━━┻━┓       ┃ ┃        
 ┃    12      ┃ ┃        
 ┃  ┏━┻┓      ┃ ┃        
 ┃  ┃  11     ┃ ┃        
 ┃  ┃ ┏┻┓     ┃ ┃        
 10 ┃ ┃ ┃     ┃ ┃        
┏┻┓ ┃ ┃ ┃     ┃ ┃        
1 2 4 0 5  3  6 9  7  8  

id      flags   population      time    metadata
0       1       0       0.00000000000000
1       1       0       0.00000000000000
2       1       0       0.00000000000000
3       1       0       0.00000000000000
4       1       0       0.00000000000000
5       1       0       0.00000000000000
6       1       0       0.00000000000000
7       1       0       0.00000000000000
8       1       0       0.00000000000000
9       1       0       0.00000000000000
10      0       0       1.00000000000000
11      0       0       26.00000000000000
12      0       0       59.00000000000000
13      0       0       64.00000000000000
14      0       0       88.00000000000000
         18          
    ┏━━━━┻━━━━┓      
    ┃         17     
    ┃       ┏━┻━━┓   
    ┃       ┃    16  
    ┃       ┃   ┏┻┓  
    ┃       15  ┃ ┃  
    ┃     ┏━┻┓  ┃ ┃  
    ┃     ┃  14 ┃ ┃  
    ┃     ┃ ┏┻┓ ┃ ┃  
    13    ┃ ┃ ┃ ┃ ┃  
 ┏━━┻━┓   ┃ ┃ ┃ ┃ ┃  
 ┃    12  ┃ ┃ ┃ ┃ ┃  
 ┃  ┏━┻┓  ┃ ┃ ┃ ┃ ┃  
 ┃  ┃  11 ┃ ┃ ┃ ┃ ┃  
 ┃  ┃ ┏┻┓ ┃ ┃ ┃ ┃ ┃  
 10 ┃ ┃ ┃ ┃ ┃ ┃ ┃ ┃  
┏┻┓ ┃ ┃ ┃ ┃ ┃ ┃ ┃ ┃  
1 2 4 0 5 8 6 9 3 7  

id      flags   population      time    metadata
0       1       0       0.00000000000000
1       1       0       0.00000000000000
2       1       0       0.00000000000000
3       1       0       0.00000000000000
4       1       0       0.00000000000000
5       1       0       0.00000000000000
6       1       0       0.00000000000000
7       1       0       0.00000000000000
8       1       0       0.00000000000000
9       1       0       0.00000000000000
10      0       0       1.00000000000000
11      0       0       26.00000000000000
12      0       0       59.00000000000000
13      0       0       64.00000000000000
14      0       0       88.00000000000000
15      0       0       336.82403893909088
16      0       0       1650.08168020564153
17      0       0       1826.65479428849403
18      0       0       3236.30194595487910
```

The API here using the ``simulator_factory`` isn't ideal, but it'll do for now. We can come up with some more elegant way of parameterising this later.

The main change here is that we can now switch around models as we see fit. When you change a model, all the internal times and rates are first converted back into generations, and then converted into model time for the new model. This can all be handled generically.

The other change I've made is the make each individual population a fraction of the model population size. (We compute N by multipling the population size by the model population size, and rounding the result.) This has the really nice property of making it the same as the coalescent models, and so all the demographic events and population size growth rates and so on should just work now. Yay!

There's some more work to be done here in unscaling and rescaling the various rates and times for demographic events, but I thought I'd share this much with you so you can get on with testing it. The hybrid model should work just fine now.

I'll get back to finish this up when I get a chance --- I don't want to merge it until there's a lot of testing on the Python side.

**Comment by DomNelson** [comment]
Awesome, looks great! Ok if I understand this properly, it's ready for testing within simple scenarios with no demographic changes or migration (basically one population with constant growth rate)?

**Comment by jeromekelleher** [comment]
Yeah, it'll work fine for a single pop with 0 growth rate and no demographic events. I need to go through the more complicated stuff to make sure the rates are all set properly.

**Comment by jeromekelleher** [comment]
I've updated this a bit more to add support for demographic events and so on. I've changed the high-level interface so that the model changing is done using a DemographicEvent (see the test in test_models.py).

I've also made a start on adding the statistical checks for the DTWF into verification.py. Looks like we have a bit of work to do here...

tmrca

I'll add some more testing to this PR when I get a chance.

**Comment by jeromekelleher** [comment]
I've done some more work to add tests here @DomNelson and get the frameworks in place for supporting everything in the DTWF that we do in the standard coalescent. I think we have a ways to go before everything works properly, but there's a clear path towards that. Here's what I think we need to do:

1. Figure out why we're a factor of 2 out on the TMRCA for a single locus. To see this, run ``python verification.py dtwf_vs_coalescent_single_locus``. Interestingly, we seem to be OK when we add some recombination in (see the ``dtwf_vs_coalescent_low_recombination`` verification check).
2. Add more checks to verification.py checking various regimes (and perhaps more statistics?). We want to be sure that the basic DTWF gives the same results as the coalescent when N is reasonably large and n << N.
3. Add verification checks for the various demographic events in the DTWF, and make sure these work well in corner cases. (Also need some C unit test coverage here.)
4. Make sure ancient sampling works well (I've added some tests for this already, but some are failing; see ``HistoricalSamplingMixin`` and ``TestHistoricalSamplingWrightFisher``. I think this a good model for getting coverage on the DTWF without duplicating a whole pile of tests.)
5. Write some tests for the ModelChangeEvent to make sure model changes work OK.
6. Do some performance work---I think there's probably some shortcuts we can take if we think about it a bit more.

We can either keep working on this PR, or create these items as issues and work on them separately. It's probably better to do the latter so we this diff doesn't grow too big. If you agree with merging this much, it would be good if you could review to make sure I haven't introduced any errors. What do you think?

**Comment by DomNelson** [comment]
This is really great @jeromekelleher to have all this set up. I'm not surprised there are a few issues to sort out, and this will help to identify where they're coming from. I'm all for splitting into more focused PRs - I'll review the changes so far and keep you updated.

File: lib/msprime.c [hunk 1]
```diff
@@ -2346,16 +2348,73 @@ msp_dtwf_generation(msp_t *self)
...
         self->time++;
         ret = msp_dtwf_generation(self);
         if (ret != 0) {
             goto out;
         }
+
+        /* TODO Need to reason more carefully about what order these events happen in!
+         * Do migrations come before or after the actual breeding? Given a sampling
+         * event at time t, does this happen before or after the breeding for time
+         * t? Same for all demographic events */
```
[the remark this answers is not in the document]
Comment (jeromekelleher) [reply lib/msprime.c]: OK, sounds good. We just need to be careful that we document clearly what we're doing I think.

File: lib/msprime.c [hunk 2]
```diff
@@ -2346,16 +2348,73 @@ msp_dtwf_generation(msp_t *self)
...
+    unsigned long events = 0;
+    int mig_source_pop, mig_dest_pop;
+    sampling_event_t *se;
+    double mu;
+    uint32_t j, k, i, num_migrations, n;
 
-    while (msp_get_num_ancestors(self) > 0 && self->time < max_time) {
+    while (msp_get_num_ancestors(self) > 0
+            && self->time < max_time && events < max_events) {
+        events++;
```
[the remark this answers is not in the document]
Comment (jeromekelleher) [reply lib/msprime.c]: This isn't really for counting events as such, it's just for roughly chunking up the work that the simulator does. It's really only there for the Python interface --- without this, you wouldn't be able to interrupt a simulation using CTRL-C, as we'd never return to the interpreter. See [here](https://github.com/jeromekelleher/msprime/blob/master/_msprimemodule.c#L8085) for how it's used.

**Comment by jeromekelleher** [comment]
OK, great, let's merge this so. I'll open a bunch of issues and assign them to one of us.
